### PR TITLE
Migrates to Clipanion 3

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

Clipanion 2 doesn't automatically infers the CLI option types, requiring to write more code and risking some types to become out of date. Additionally, it depends on the decorator syntax, which is a bit too magical.

**How did you fix it?**

Clipanion 3 uses a slightly different syntax that automatically infers types from the option "decorators".

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-essentials/sources/commands/run.ts [hunk 1]
```diff
@@ -60,7 +34,34 @@ export default class RunCommand extends BaseCommand {
...
+  // Some tools (for example text editors) want to call the real binaries, not
+  // what their users might have remapped them to in their `scripts` field.
+  binariesOnly = Option.Boolean(`-B,--binaries-only`, false, {hidden: true});
+
+  // The v1 used to print the Yarn version header when using "yarn run", which
+  // was messing with the output of things like `--version` & co. We don't do
+  // this anymore, but many workflows use `yarn run --silent` to make sure that
+  // they don't get this header, and it makes sense to support it as well (even
+  // if it's a no-op in our case).
+  silebt = Option.Boolean(`--silent`, {hidden: true});
```
Comment (paul-soporan) [inline packages/plugin-essentials/sources/commands/run.ts:60]:
```suggestion
  silent = Option.Boolean(`--silent`, {hidden: true});
```

File: packages/plugin-workspace-tools/sources/commands/foreach.ts [hunk 2]
```diff
@@ -56,53 +56,9 @@ const getWorkspaceDependenciesRecursive = (rootWorkspace: Workspace, project: Pr
...
-  @Command.Boolean(`--no-private`, {description: `Avoid running the command on private workspaces`})
-  publicOnly: boolean = false;
-
-  static schema = yup.object().shape({
-    jobs: yup.number().min(2),
-    parallel: yup.boolean().when(`jobs`, {
-      is: (val: number) => val > 1,
-      then: yup.boolean().oneOf([true], `--parallel must be set when using --jobs`),
-      otherwise: yup.boolean(),
-    }),
```
Comment (paul-soporan) [inline packages/plugin-workspace-tools/sources/commands/foreach.ts:104]: Did you forget to port this part of the schema?

Comment (arcanis) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts:104]: Hm yep 🤔

Comment (arcanis) [reply packages/plugin-workspace-tools/sources/commands/foreach.ts:104]: I'll skip it for now - `when` isn't implemented in Typanion yet, and there are some subtleties around how it should behave with regard to type inference.

File: packages/plugin-essentials/sources/commands/config/set.ts [hunk 3]
```diff
@@ -1,25 +1,13 @@
...
 import {Configuration, StreamReport, MessageName, miscUtils} from '@yarnpkg/core';
-import {Command, Usage, UsageError}                          from 'clipanion';
+import {Command, Option, Usage, UsageError}                  from 'clipanion';
 import cloneDeep                                             from 'lodash/cloneDeep';
 import getPath                                               from 'lodash/get';
 import setPath                                               from 'lodash/set';
 import {inspect}                                             from 'util';
 
 // eslint-disable-next-line arca/no-default-export
 export default class ConfigSetCommand extends BaseCommand {
```
Comment (merceyz) [inline packages/plugin-essentials/sources/commands/config/set.ts:10]: Seems to lack a path now

**Comment by MeLlamoPablo** [comment]
Hello @arcanis 

This seems to have broken the `plugin-interactive-tools` build (I don't know for any other plugins). When I build it using `yarn build:plugin-interactive-tools`, then import it into another project, I get:

```
Usage Error: Couldn't find a script named "upgrade-interactive", but a matching command can be found in the interactive-tools plugin. You can install it with "yarn plugin import interactive-tools".

$ yarn run [--inspect] [--inspect-brk] <scriptName> ...
```

Because I see that the `packages/plugin-interactive-tools/bin/@yarnpkg/plugin-interactive-tools.js` hasn't been updated, I wanted to ask: is this intended?

I did not open an issue because I didn't want to go through all the Sherlock bureaucracy just to find out that this was intended and will be fixed soon. If you tell me this is unintentional I will open an issue through the proper channels.

**Comment by arcanis** [comment]
> Because I see that the `packages/plugin-interactive-tools/bin/@yarnpkg/plugin-interactive-tools.js` hasn't been updated, I wanted to ask: is this intended?

Yep. We're currently in-between majors so there hasn't been a formal release for some time, and the files in `bin` are only updated during each releases. The next one, which I'd like to release soonish (a few weeks?), will likely be `3.0.0-rc.1`.

**Comment by merceyz** [comment]
> When I build it using yarn build:plugin-interactive-tools, then import it into another project, I get:

Keep in mind that updates `packages\plugin-interactive-tools\bundles\@yarnpkg\plugin-interactive-tools.js`

**Comment by MeLlamoPablo** [comment]
Thanks @arcanis. I was asking because I wanted a send a feature request PR for supporting "ignore lists" on `yarn upgrade-interactive`.

What would you recommend to me? Should I develop using the Clipanion 2 version as base and deal with conflicts later? Or should I wait until the next release?

**Comment by arcanis** [comment]
You can already use the master build by running `yarn set version from sources` - you should then be able to use `yarn plugin import from sources --branch <pr #>` to use your PR inside your own repo.

**Pull request merged**